### PR TITLE
Mavstation addition

### PR DESCRIPTION
**Pull request by kbrtny**
This is all of the additions to get the Mavstation board running based off of this branch.

The board boots, starts a NSH terminal over USART1 (GPS port), then automatically starts the mavstation code which switches the Radio from USART2 to the Raspberry Pi Header.
The select button can be used to switch the Radio port back.
The blue LED indicates the state of the switch, on=RasPi, off=Mavstation.

### CODE REVIEW

File: src/drivers/boards/mavstation/module.mk [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+SRCS		= mavstation_init.c \
+		  mavstation_pwm_servo.c \
+		  mavstation_usb.c \
+		  mavstation_led.c
+
+ENABLE_CXXINITIALIZE=$(call check_nuttx_config ,"CONFIG_HAVE_CXX 1", $(NUTTX_CONFIG_H))
+ENABLE_CXXINITIALIZE+=$(call check_nuttx_config ,"CONFIG_HAVE_CXXINITIALIZE 1", $(NUTTX_CONFIG_H))
+ifeq ("$(ENABLE_CXXINITIALIZE)",$(nuttx_config_2true))
+SRCS		 +=	   ../../../modules/systemlib/up_cxxinitialize.c
```
Comment (LorenzMeier) [inline src/drivers/boards/mavstation/module.mk]: Does this work as intended? We had issues with this construct (the config extraction) earlier.

File: Images/mavstation.prototype [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+    "magic": "MAVSTATION", 
+    "description": "Firmware for the Mavstation board", 
+    "image": "", 
+    "build_time": 0, 
+    "summary": "MAVSTATION",
+    "version": "0.1",
+    "image_size": 0,
+    "git_identity": "",
+    "board_revision": 0
+}
```
Comment (LorenzMeier) [inline Images/mavstation.prototype]: Missing line break at EOF

File: makefiles/toolchain_gnu-arm-eabi.mk [hunk 3]
```diff
@@ -60,7 +60,7 @@ endif
 
 # XXX this is pulled pretty directly from the fmu Make.defs - needs cleanup
 
-MAXOPTIMIZATION		 ?= -O3
+MAXOPTIMIZATION		 ?= -Os
```
Comment (LorenzMeier) [inline makefiles/toolchain_gnu-arm-eabi.mk]: We can't change the optimization level in general. Please set the optimization options per module in the module.mk. NuttX is already built with -Os

File: makefiles/config_mavstation_default.mk [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+ROMFS_ROOT	 = $(PX4_BASE)/ROMFS/mavstation_common
+
+#
+# Board support modules
+#
+MODULES		+= drivers/device
+MODULES		+= drivers/stm32
+MODULES		+= drivers/boards/mavstation
+MODULES		+= modules/mavstation_firmware
+#MODULES		+= modules/uORB
```
Comment (LorenzMeier) [inline makefiles/config_mavstation_default.mk]: Please avoid commented lines in final code / makefiles.

File: ROMFS/mavstation_common/init.d/rcS [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+
+mavstation 
+
+#if sercon
+#then
+#	echo "[i] USB interface connected"
+
+	# Try to get an USB console
+#	nshterm /dev/ttyACM0 &
+#fi
```
Comment (LorenzMeier) [inline ROMFS/mavstation_common/init.d/rcS]: Commented lines, missing line break.

File: nuttx-configs/mavstation/nsh/Make.defs [hunk 6]
```diff
@@ -0,0 +1,173 @@
+############################################################################
+# configs/px4io-v1/nsh/Make.defs
```
Comment (LorenzMeier) [inline nuttx-configs/mavstation/nsh/Make.defs]: comment lies.

File: nuttx-configs/mavstation/src/Makefile [hunk 7]
```diff
@@ -0,0 +1,84 @@
+############################################################################
+# configs/stm3210e-eval/src/Makefile
```
Comment (LorenzMeier) [inline nuttx-configs/mavstation/src/Makefile]: comment lies

File: nuttx-configs/mavstation/src/README.txt [hunk 8]
```diff
@@ -0,0 +1 @@
+This directory contains drivers unique to the STMicro STM3210E-EVAL development board.
```
Comment (LorenzMeier) [inline nuttx-configs/mavstation/src/README.txt]: Comment lies

File: src/drivers/boards/mavstation/module.mk (hunk 1 above)
Comment (kbrtny) [inline src/drivers/boards/mavstation/module.mk]: It doesn't not work... I had issues with this earlier but this snippet seems to work.  I didn't see the changes you made to revert back from this for the other configurations.  I can update the format based on the other configurations based on your commits from 5 days ago.

File: src/drivers/boards/mavstation/board_config.h [hunk 9]
```diff
@@ -0,0 +1,86 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2012-2014 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/drivers/boards/mavstation/board_config.h]: 2015

File: src/drivers/boards/mavstation/mavstation_init.c [hunk 10]
```diff
@@ -0,0 +1,148 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2012 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/drivers/boards/mavstation/mavstation_init.c]: (c) 2012-2015 (note the lowercase c)

File: src/drivers/boards/mavstation/mavstation_init.c [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+#	if defined(CONFIG_EXAMPLES_NSH_CXXINITIALIZE)
+#  		error CONFIG_EXAMPLES_NSH_CXXINITIALIZE Must not be defined! Use CONFIG_HAVE_CXX and CONFIG_HAVE_CXXINITIALIZE.
+#	endif
+
+#else
+#  error platform is dependent on c++ both CONFIG_HAVE_CXX and CONFIG_HAVE_CXXINITIALIZE must be defined.
+#endif
+	stm32_usbinitialize();
+
+//	message("[boot] Successfully initialized mavstation gpios\n");
```
Comment (LorenzMeier) [inline src/drivers/boards/mavstation/mavstation_init.c]: commented code

File: src/drivers/boards/mavstation/mavstation_led.c [hunk 12]
```diff
@@ -0,0 +1,127 @@
+/****************************************************************************
+ *
+ *   Copyright (c) 2013 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/drivers/boards/mavstation/mavstation_led.c]: 2013-2015

File: src/drivers/boards/mavstation/mavstation_pwm_servo.c [hunk 13]
```diff
@@ -0,0 +1,93 @@
+/****************************************************************************
+ *
+ *  Copyright (C) 2012 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/drivers/boards/mavstation/mavstation_pwm_servo.c]: 2012-2015 and lowercase c

File: src/examples/px4_simple_app/px4_simple_app.c [hunk 14]
```diff
@@ -75,7 +76,7 @@ int px4_simple_app_main(int argc, char *argv[])
 	for (int i = 0; i < 5; i++) {
 		/* wait for sensor update of 1 file descriptor for 1000 ms (1 second) */
 		int poll_ret = poll(fds, 1, 1000);
-	 
+
```
Comment (LorenzMeier) [inline src/examples/px4_simple_app/px4_simple_app.c]: this whitespace change is welcome.

File: src/examples/px4_simple_app/px4_simple_app.c [hunk 15]
```diff
@@ -42,6 +42,7 @@
 #include <stdio.h>
 #include <poll.h>
 #include <string.h>
+ #include <math.h>
```
Comment (LorenzMeier) [inline src/examples/px4_simple_app/px4_simple_app.c]: this whitespace change is not welcome :)

File: src/modules/mavstation_firmware/adc.c [hunk 16]
```diff
@@ -0,0 +1,170 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2012 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/adc.c]: (c) 2012-2015

File: src/modules/mavstation_firmware/appdebug.h [hunk 17]
```diff
@@ -0,0 +1,16 @@
+
+#ifndef __MAVSTATION_FIRMWARE_APPDEBUG_H__
+#define __MAVSTATION_FIRMWARE_APPDEBUG_H__
+
+# define debug(fmt, args...)	lowsyslog(LOG_DEBUG,fmt "\n", ##args)
+/*
+ * add a debug message to be printed on the console
+ */
+void isr_debug(uint8_t level, const char *fmt, ...);
+//void debug(const char* fmt, ...);
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/appdebug.h]: commented code

File: src/modules/mavstation_firmware/i2c_slave.c [hunk 18]
```diff
@@ -0,0 +1,351 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2012,2013 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/i2c_slave.c]: (c) 2012-2015

File: src/modules/mavstation_firmware/i2c_slave.c [hunk 19]
```diff
@@ -0,0 +1,351 @@
...
+#include <stm32_dma.h>
+
+#include "appdebug.h"
+#include "i2c_slave.h"
+#include "slave_registers.h"
+
+/*
+ * I2C register definitions.
+ */
+#define I2C_BASE	STM32_I2C1_BASE
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/i2c_slave.c]: This entire file looks like its duplicate to the PX4IO file and hampers maintainability. Can they be reconciled?

File: src/modules/mavstation_firmware/mavstation.c [hunk 20]
```diff
@@ -0,0 +1,165 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2012 PX4 Development Team. All rights reserved.
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/mavstation.c]: (c) 2012-2015

File: src/modules/mavstation_firmware/mavstation.c [hunk 21]
```diff
@@ -0,0 +1,165 @@
...
+	 */
+#ifdef CONFIG_ARCH_DMA
+	hrt_call_every(&serial_dma_call, 1000, 1000, (hrt_callout)stm32_serial_dma_poll, NULL);
+#endif
+
+	/* print some startup info */
+	debug("\nmavstation: starting\n");
+
+	/* configure the first 4 PWM outputs */
+	//up_pwm_servo_init(0x0f);
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/mavstation.c]: commented code

File: src/modules/mavstation_firmware/mavstation.c [hunk 22]
```diff
@@ -0,0 +1,165 @@
...
+				gpio_interface_setled(2,0);
+				MuxState=1;
+			}
+			MuxFlag=1;
+		}else if(!gpio_interface_getbtn(0))
+		{
+			MuxFlag=0;
+		}
+
+#define DEBUG_GPIOS
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/mavstation.c]: Enabled debug option

File: src/modules/mavstation_firmware/protocol.h [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ * Mavstation I2C interface notes
+ * -------------------------
+ *
+ * Register read/write operations are mapped directly to PX4IO register
+ * read/write operations.
+ *
+ */
+
+#define PX4IO_CONTROL_CHANNELS			8
```
Comment (LorenzMeier) [inline src/modules/mavstation_firmware/protocol.h]: If its using the same API as PX4IO, both should use the same files.

**Comment by LorenzMeier** [comment]
@kbrtny Thanks! Could you explain the rationale behind the code duplication on the I2C API with PX4IO? @craig3dr This looks like it will be a nightmare to maintain and almost guarantee the mavstation config will become quickly stale. It would be much better to reconcile the I2C API code between the two boards.

**Comment by kbrtny** [comment]
I'll fix up the comments in the morning.  The I2c code was in there before I started on the project and is intended to function similarly to the PX4IO board.  It's currently not in use and I'm not sure if there are plans to roll it in.

**Comment by davidbuzz** [comment]
the i2c code may have come from my earlier attempts to get the mavstation working, not sure.     it was my hope that (sometime in the future) we'd be able to have i2c comms as another channel available between the two cpu's  ( one in the RPi, and the other being the stm32f1 in the mavstation daughter board).       If it is made to work,  it could ( in theory ) be used to push new firmware to the stm32f1, or as a channel to feed I/O information back to teh RPi. ( eg the state of the pushbuttons on the mavstation, for example), or GPS data.         I'm not entirely sure it's even electrically possible at this stage, given I don't know if the hardware can do this.... so it's still unknown.   If you are worried, maybe it's best to leave the i2c code out of the mavstation for now. ?

**Comment by davidbuzz** [comment]
Also.    other than lorenz's obsession with comments, whitespace, and copyright, it looks pretty good to me.

**Comment by Craig3DRobotics** [comment]
That I2C code would have come from Pat when he was working on this project.

**Comment by LorenzMeier** [comment]
@davidbuzz You should get code review from Google (or any other valley company). I was in fact kind ;-). You wouldn't even allowed to commit this (a pre-commit hook will pre-empt you from that). An industry-standard code review would have turned out many more lines.

**Comment by LorenzMeier** [comment]
Ok, then lets get rid of the I2C stuff to avoid bit-rot.

**Comment by davidbuzz** [comment]
@Lorenz ... may I suggest you automate this sort of whitespace/copyright/eof checks...? Perhaps add a uncrustify or mogrify as part of the Jenkins CI ...?  It's surely easier to code a fix once than to manually check for this sort of thing constantly. ..

**Comment by LorenzMeier** [comment]
Didn't actually intent to close this. @davidbuzz We have the tools in Tools/fix_code_style.sh but its a delicate topic. If we enforce too strict, we will just start to kill off PRs in the first place. There is a pending pull request which would do strict checking using Travis. But we probably need something softer, and we need to convert / format the parts of the codebase which deviate (not terribly much, actually).

That tool doesn't touch comments, though, as that's a rather semantic insight.

@kbrtny Could you run the fix code style scripts on all files here as well?

**Comment by kbrtny** [comment]
@LorenzMeier That should take care of it.

**Comment by LorenzMeier** [comment]
I still don't understand the I2C bits in this PR. What does talk via I2C to this board? Or is this copy & paste cruft from the PX4IO driver?

**Comment by davidbuzz** [comment]
@lorenz ... the i2c features should in the future allow the RPi and the stm32 to communicate between themselves but that part of the code isn't written yet. The px4io code is the best current candidate we have so far that should be a good basis to start from . It's not important right now but we hope it will be in the future.

**Comment by LorenzMeier** [comment]
If the I2C code is completely untested I would prefer to not pull it in, because in general we try to provide a good starting point to developers picking up loose ends. So what state is it in? If its just a ~1 year old copy of the IO code someone would be better off doing a new copy when the time comes to implement I2C, as else the stale copy will miss all the bug fixes and improvements done since then in the IO codebase.

**Comment by LorenzMeier** [comment]
I rebased / squashed it for better history maintainability into a single commit and added it onto nuttx_next, which is the current nuttx work rebased on master:
https://github.com/PX4/Firmware/tree/nuttx_next

Thanks!

**Pull request closed without merging**